liberapay/security/authentication.py

```diff
@@ -57,6 +63,41 @@ def sign_in_with_form_data(body, state):
             )
             if not p:
                 state['log-in.error'] = _("Bad username or password.")
+            else:
+                last_password_check = p.get_last_event_of_type('password-check')
+                if (not last_password_check) or utcnow() - last_password_check.ts > timedelta(days=30):
+                    passhash = sha1(password.encode("utf-8")).hexdigest().upper()
+                    passhash_short = passhash[:5]
+                    URL = "https://api.pwnedpasswords.com/range/"+passhash_short
+
+                    r = requests.get(url=URL)
+                    count = 0
+                    for line in r.text.split("\n"):
+                        suffix = line.split(":")[0]
+                        if passhash_short + suffix == passhash:
+                            count = int(line.split(":")[1].strip())
+                    strength, improvements = passwordmeter.test(password)
+                    if strength < 0.3:
+                        password_check_message = _(
+                            "We have detected that your current password is weak. "
+                            "We recommend you to choose a combination of letters, numbers and symbols."
+                        )
+                        p.notify('password_warning', email=False, message=password_check_message)
+                    elif count > 500:
+                        password_check_message = _(
+                            "We have detected that your current password is commonly used as a password. "
+                            "You are recommended to change your password."
+                        )
+                        p.notify('password_warning', email=False, message=password_check_message)
+                    elif count > 0:
+                        password_check_message = _(
+                            "We have detected that your current password has been compromised in leaks "
+                            "on the internet before. "
+                            "You are recommended to change your password."
+                        )
+                        p.notify('password_warning', email=False, message=password_check_message)
+                    p.add_event(website.db, 'password-check', None)
```

Comment:
This should be moved to a separate function, and the call to that new function should be wrapped to report and mask errors, like this:

```python
try:
    check_password(password)
except Exception as e:
    website.tell_sentry(e, state)
```